designs/cache_options.md

```diff
@@ -0,0 +1,170 @@
+Cache Options
+===================
+
+This document describes how we imagine the cache options to look in
+the future.
+
+## Goals
+
+* Align everyone on what settings on the cache we want to support and
+  their configuration surface
+* Ensure that we support both complicated cache setups and provide an
+  intuitive configuration UX
+
+## Non-Goals
+
+* Describe the design and implementation of the cache itself.
+  The assumption is that the most granular level we will end up with is
+  "per-object multiple namespaces with distinct selectors" and that this
+  can be implemented using a "meta cache" that delegates per object and by
+  extending the current multi-namespace cache
+* Outline any kind of timeline for when these settings will be implemented.
+  Implementation will happen gradually over time whenever someone steps up
+  to do the actual work
+
+## Proposal
+
+
+```
+type CacheSetting struct {
+  // LabelSelector specifies a label selector. A nil value allows to
+  // default this.
+  LabelSelector labels.Selector
+
+  // FieldSelector specifics a field selector. A nil value allows to
+  // default this.
+  FieldSelector fields.Selector
+
+  // Transform specifies a transform func. A nil value allows to default
+  // this.
+  Transform     toolscache.TransformFunc
+
+  // UnsafeDisableDeepCopy specifies if List and Get requests against the
+  // cache should not DeepCopy. A nil value allows to default this.
+  UnsafeDisableDeepCopy *bool
+}
+
+
+type ByObject struct {
+  // Namespaces maps a namespace name to cache setting. If set, only the
+  // namespaces in this map will be cached.
+  //
+  // A nil value in this map means "fall through to the ByObject CacheSetting"
+  //
+  // An empty value in this map means "No Selectors".
+  //
+  // It is possible to have specific CacheSettings for just some namespaces
+  // but cache all namespaces by using the empty string as the map key for
+  // "all namespaces". This wil then include all namespaces that do not have
+  // a more specific setting.
+  //
+  // A nil map allows to default this to the cache's DefaultNamespaces setting.
+  // An empty map prevents this.
+  //
+  // This must be unset for cluster-scoped objects.
+  Namespaces map[string]*CacheSetting
+
+  // CacheSettings will be used if Namespaces is empty or for entries in
+  // Namespaces that have a nil value.
+  //
+  // If nil, this will be defaulted to the cache's DefaultLabelSelector,
+  // DefaultFieldSelector, and DefaultTransform.
+  CacheSettings *CacheSetting
+}
+
+type Options struct {
```

Comment:
1. So where is the other fields like `Scheme`, `SyncPeriod`? They should remain in Options, right?
2. Will the `Namespaces []string` be replaced by the new `DefaultNamespaces map[string]*CacheSetting`?
3. Is the `DefaultTransform` field removed?

Reply:
1. Yes, I guess I didn't include them because they are global options, will add
2. Yes
3. No, that is an oversight, will add

Reply:
Added the missing options, ptal